File: examples/gpt2_generate.py
#!/usr/bin/env python3
# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
#                              (Skoltech), Russia. All rights reserved.
#                2023-present Artificial Intelligence Research Institute
#                              (AIRI), Russia. All rights reserved.
#
# NNTile is software framework for fast training of big neural networks on
# distributed-memory heterogeneous systems based on StarPU runtime system.
#
# @file examples/gpt2_generate.py
# Convert a HuggingFace GPT-2 checkpoint to NNTile format for C++ inference.
#
# Usage:
#   python gpt2_generate.py \
#       --model gpt2 \
#       --output-dir /tmp/nntile_gpt2 \
#       --prompt "The capital of France is"
#
# Then run the C++ binary:
#   ./gpt2_generate \
#       --config /tmp/nntile_gpt2/config.json \
#       --weights /tmp/nntile_gpt2/weights.safetensors \
#       --prompt-ids "$(cat /tmp/nntile_gpt2/prompt_ids.txt)" \
#       --max-tokens 32
#
# @version 1.1.0

"""Convert HuggingFace GPT-2 checkpoint to NNTile format and tokenize a prompt.

Produces:
  - weights.safetensors  NNTile-layout weight file
  - config.json  Model configuration readable by the C++ example
  - prompt_ids.txt  Comma-separated token IDs for the prompt (if --prompt supplied)
"""

from __future__ import annotations

import argparse
import json
import struct
import sys
from pathlib import Path
from typing import Callable, Sequence

import numpy as np
import torch
from huggingface_hub import snapshot_download
from safetensors import safe_open
from transformers import AutoConfig, AutoTokenizer


def fortran_order(arr: np.ndarray) -> np.ndarray:
    """Return C-contiguous array whose flat bytes equal NNTile column-major."""
    a = np.asarray(arr, dtype=np.float32)
    return a.ravel("F").reshape(a.shape)


def _write_safetensors_streaming(
    path: str | Path,
    specs: Sequence[tuple[str, tuple[int, ...]]],
    get_data: Callable[[str], np.ndarray],
) -> None:
    """Write a safetensors file one tensor at a time."""
    header: dict[str, dict] = {}
    offset = 0
    entry_order: list[tuple[str, int]] = []
    for name, shape in specs:
        nbytes = int(np.prod(shape)) * 4
        header[name] = {
            "dtype": "F32",
            "shape": list(shape),
            "data_offsets": [offset, offset + nbytes],
        }
        entry_order.append((name, nbytes))
        offset += nbytes

    header_bytes = json.dumps(header, separators=(",", ":")).encode("utf-8")
    pad = (8 - len(header_bytes) % 8) % 8
    header_bytes += b" " * pad

    with open(path, "wb") as f:
        f.write(struct.pack("<Q", len(header_bytes)))
        f.write(header_bytes)
        for name, nbytes in entry_order:
            arr = get_data(name)
            raw = arr.tobytes()
            assert len(raw) == nbytes, f"{name}: expected {nbytes} bytes, got {len(raw)}"
            f.write(raw)
            del arr, raw


def _output_specs(config) -> list[tuple[str, tuple[int, ...]]]:
    """Return (nntile_name, shape) for every output tensor."""
    H = config.n_embd
    V = config.vocab_size
    n_layer = config.n_layer
    n_head = config.n_head
    n_inner = getattr(config, "n_inner", None) or 4 * H
    head_size = H // n_head

    specs: list[tuple[str, tuple[int, ...]]] = []

    specs.append(("model.transformer.wte.vocab", (H, V)))
    specs.append(("model.transformer.wpe.vocab", (H, config.n_positions)))

    for i in range(n_layer):
        p = f"model.transformer.h_{i}"
        specs.append((f"{p}.ln_1.gamma", (H,)))
        specs.append((f"{p}.ln_2.gamma", (H,)))

        specs.append((f"{p}.attn.q_weight", (n_head, head_size, H)))
        specs.append((f"{p}.attn.k_weight", (n_head, head_size, H)))
        specs.append((f"{p}.attn.v_weight", (n_head, head_size, H)))
        specs.append((f"{p}.attn.o_weight", (H, n_head, head_size)))

        specs.append((f"{p}.mlp.fc1.weight", (H, n_inner)))
        specs.append((f"{p}.mlp.fc2.weight", (n_inner, H)))

    specs.append(("model.transformer.ln_f.gamma", (H,)))
    specs.append(("model.lm_head.weight", (H, V)))

    return specs


def _make_converter(
    config,
    hf_get: Callable[[str], np.ndarray],
    tensor_keys: set,
) -> Callable[[str], np.ndarray]:
    """Return a function that converts a single NNTile tensor on demand."""
    H = config.n_embd
    n_head = config.n_head
    head_size = H // n_head
    n_inner = getattr(config, "n_inner", None) or 4 * H

    def convert(name: str) -> np.ndarray:
        if name == "model.transformer.wte.vocab":
            return fortran_order(hf_get("wte.weight").T)

        if name == "model.transformer.wpe.vocab":
            return fortran_order(hf_get("wpe.weight").T)

        if name == "model.transformer.ln_f.gamma":
            return fortran_order(hf_get("ln_f.weight"))

        if name == "model.lm_head.weight":
            # GPT2LMHeadModel ties lm_head to wte; use wte if lm_head missing
            if "lm_head.weight" in tensor_keys:
                return fortran_order(hf_get("lm_head.weight").T)
            return fortran_order(hf_get("wte.weight").T)

        parts = name.split(".")
        layer_idx = int(parts[2].split("_", 1)[1])
        rest = ".".join(parts[3:])
        hp = f"h.{layer_idx}"

        if rest == "ln_1.gamma":
            return fortran_order(hf_get(f"{hp}.ln_1.weight"))
        if rest == "ln_2.gamma":
            return fortran_order(hf_get(f"{hp}.ln_2.weight"))

        if rest == "attn.q_weight":
            c_attn = hf_get(f"{hp}.attn.c_attn.weight")
            q = c_attn[:, :H].T.reshape(n_head, head_size, H)
            return fortran_order(q)
        if rest == "attn.k_weight":
            c_attn = hf_get(f"{hp}.attn.c_attn.weight")
            k = c_attn[:, H:2*H].T.reshape(n_head, head_size, H)
            return fortran_order(k)
        if rest == "attn.v_weight":
            c_attn = hf_get(f"{hp}.attn.c_attn.weight")
            v = c_attn[:, 2*H:3*H].T.reshape(n_head, head_size, H)
            return fortran_order(v)
        if rest == "attn.o_weight":
            o = hf_get(f"{hp}.attn.c_proj.weight").T.reshape(H, n_head, head_size)
            return fortran_order(o)

        if rest == "mlp.fc1.weight":
            return fortran_order(hf_get(f"{hp}.mlp.c_fc.weight").T)
        if rest == "mlp.fc2.weight":
            return fortran_order(hf_get(f"{hp}.mlp.c_proj.weight").T)

        raise ValueError(f"Unknown NNTile tensor: {name}")

    return convert


def _load_tokenizer(*sources: str):
    """Try to load a tokenizer from multiple sources."""
    for src in sources:
        if not src:
            continue
        try:
            return AutoTokenizer.from_pretrained(src, use_fast=False)
        except Exception:
            pass
    return None


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Convert HF GPT-2 checkpoint → NNTile format",
    )
    parser.add_argument(
        "--model", required=True,
        help="HuggingFace model name (e.g. gpt2) or local path",
    )
    parser.add_argument(
        "--output-dir", required=True,
        help="Directory where NNTile files will be written",
    )
    parser.add_argument(
        "--prompt", default=None,
        help="Optional text prompt to tokenize",
    )
    args = parser.parse_args()

    model_id = args.model

    if Path(model_id).is_dir():
        model_dir = Path(model_id)
    else:
        print(f"Downloading {model_id} to HF cache ...")
        model_dir = Path(snapshot_download(model_id))
        print(f"Snapshot: {model_dir}")

    out_dir = Path(args.output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)

    config = AutoConfig.from_pretrained(str(model_dir))
    print(f"Model: {config.model_type}  hidden={config.n_embd}  "
          f"layers={config.n_layer}  vocab={config.vocab_size}")

    n_inner = getattr(config, "n_inner", None) or 4 * config.n_embd
    nntile_config = {
        "vocab_size": config.vocab_size,
        "n_embd": config.n_embd,
        "n_inner": n_inner,
        "n_layer": config.n_layer,
        "n_head": config.n_head,
        "n_positions": config.n_positions,
        "layer_norm_epsilon": getattr(config, "layer_norm_epsilon", 1e-5),
        "eos_token_id": getattr(config, "eos_token_id", 50256),
        "bos_token_id": getattr(config, "bos_token_id", 50256),
    }
    config_path = out_dir / "config.json"
    config_path.write_text(json.dumps(nntile_config, indent=2) + "\n")
    print(f"Wrote {config_path}")

    st_files = sorted(model_dir.glob("*.safetensors"))
    if not st_files:
        print(f"ERROR: no *.safetensors files found in {model_dir}", file=sys.stderr)
        return 1

    handles: list = []
    tensor_to_handle: dict[str, object] = {}
    for sf in st_files:
        h = safe_open(str(sf), framework="pt")
        handles.append(h)
        for key in h.keys():
            tensor_to_handle[key] = h

    def hf_get(name: str) -> np.ndarray:
        return tensor_to_handle[name].get_tensor(name).to(torch.float32).numpy()

    tensor_keys = set(tensor_to_handle.keys())
    specs = _output_specs(config)
    converter = _make_converter(config, hf_get, tensor_keys)

    weights_path = out_dir / "weights.safetensors"
    print(f"Converting {len(specs)} tensors (streaming) ...")
    _write_safetensors_streaming(weights_path, specs, converter)
    print(f"Wrote {weights_path}")

    del handles, tensor_to_handle

    if args.prompt is not None:
        tokenizer = _load_tokenizer(str(model_dir), model_id)
        if tokenizer is None:
            print("WARNING: could not load tokenizer; skipping prompt tokenization.",
                  file=sys.stderr)
        else:
            ids = tokenizer.encode(args.prompt, add_special_tokens=True)
            ids_str = ",".join(str(i) for i in ids)
            ids_path = out_dir / "prompt_ids.txt"
            ids_path.write_text(ids_str + "\n")
            print(f"Prompt ({len(ids)} tokens): {ids}")
            print(f"Wrote {ids_path}")

    print("Done.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/gpt2_generate.cc
/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
 *                              (Skoltech), Russia. All rights reserved.
 *                 2023-present Artificial Intelligence Research Institute
 *                              (AIRI), Russia. All rights reserved.
 *
 * NNTile is software framework for fast training of big neural networks on
 * distributed-memory heterogeneous systems based on StarPU runtime system.
 *
 * @file examples/gpt2_generate.cc
 * Autoregressive text generation with a GPT-2 causal LM.
 *
 * Workflow (two-step):
 *
 *   # Step 1 – Convert HF checkpoint and tokenize the prompt (Python):
 *   python examples/gpt2_generate.py \
 *       --model gpt2 \
 *       --output-dir /tmp/nntile_gpt2 \
 *       --prompt "The meaning of life is"
 *
 *   # Step 2 – Run autoregressive generation (this binary):
 *   ./gpt2_generate \
 *       --config /tmp/nntile_gpt2/config.json \
 *       --weights /tmp/nntile_gpt2/weights.safetensors \
 *       --prompt-ids "$(cat /tmp/nntile_gpt2/prompt_ids.txt)" \
 *       --max-tokens 32
 *
 * Alternatively, let the binary call the Python script automatically:
 *   ./gpt2_generate \
 *       --model gpt2 \
 *       --prompt "The meaning of life is" \
 *       --max-tokens 32
 *
 * @version 1.1.0
 * */

#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <sys/wait.h>
#include <unistd.h>
#include <vector>

#include <nntile.hh>
#include <nntile/graph/io/safetensors.hh>
#include <nntile/graph/model/gpt2/gpt2_causal.hh>
#include <nntile/graph/model/gpt2/gpt2_config.hh>
#include <nlohmann/json.hpp>

using namespace nntile;
using namespace nntile::graph;
using namespace nntile::model::gpt2;
using json = nlohmann::json;

// ── CLI helpers ──────────────────────────────────────────────────────────

struct Args
{
    std::string config_path;
    std::string weights_path;
    std::string prompt_ids_str;
    std::string model;
    std::string prompt;
    std::string output_dir = "/tmp/nntile_gpt2";
    int max_tokens = 32;
};

static Args parse_args(int argc, char** argv)
{
    Args a;
    for(int i = 1; i < argc; ++i)
    {
        std::string arg = argv[i];
        if(arg == "--config" && i + 1 < argc)
            a.config_path = argv[++i];
        else if(arg == "--weights" && i + 1 < argc)
            a.weights_path = argv[++i];
        else if(arg == "--prompt-ids" && i + 1 < argc)
            a.prompt_ids_str = argv[++i];
        else if((arg == "--model" || arg == "--model-dir") && i + 1 < argc)
            a.model = argv[++i];
        else if(arg == "--prompt" && i + 1 < argc)
            a.prompt = argv[++i];
        else if(arg == "--output-dir" && i + 1 < argc)
            a.output_dir = argv[++i];
        else if(arg == "--max-tokens" && i + 1 < argc)
            a.max_tokens = std::stoi(argv[++i]);
        else if(arg == "--help" || arg == "-h")
        {
            std::cout
                << "Usage: " << argv[0] << " [options]\n\n"
                << "Option A – provide pre-converted files:\n"
                << "  --config <path>       config.json (from gpt2_generate.py)\n"
                << "  --weights <path>      weights.safetensors\n"
                << "  --prompt-ids <ids>    comma-separated token IDs\n\n"
                << "Option B – let the binary call Python automatically:\n"
                << "  --model <name|path>   HF model name or local path\n"
                << "  --prompt <text>       text prompt\n"
                << "  --output-dir <path>   scratch directory "
                   "(default: /tmp/nntile_gpt2)\n\n"
                << "Common:\n"
                << "  --max-tokens <N>      tokens to generate (default: 32)\n";
            std::exit(0);
        }
    }
    return a;
}

static std::vector<std::int64_t> parse_ids(const std::string& s)
{
    std::string data = s;
    {
        std::ifstream f(s);
        if(f.good())
        {
            std::string contents((std::istreambuf_iterator<char>(f)),
                                  std::istreambuf_iterator<char>());
            if(!contents.empty())
                data = contents;
        }
    }

    std::vector<std::int64_t> ids;
    std::istringstream ss(data);
    std::string tok;
    while(std::getline(ss, tok, ','))
    {
        while(!tok.empty() && (tok.back() == '\n' || tok.back() == '\r'))
            tok.pop_back();
        if(!tok.empty())
            ids.push_back(std::stoll(tok));
    }
    return ids;
}

// ── Config loader ────────────────────────────────────────────────────────

// Extract int from JSON; accepts both number and string (e.g. "32000").
static int config_get_int(const json& j, const char* key, int default_val)
{
    if(!j.contains(key))
        return default_val;
    const auto& v = j[key];
    if(v.is_number_integer())
        return v.get<int>();
    if(v.is_number_float())
        return static_cast<int>(v.get<double>());
    if(v.is_string())
        return std::stoi(v.get<std::string>());
    throw std::runtime_error(std::string("config: '") + key +
                            "' must be int or string, got " + v.type_name());
}

// Extract float from JSON; accepts both number and string.
static float config_get_float(const json& j, const char* key, float default_val)
{
    if(!j.contains(key))
        return default_val;
    const auto& v = j[key];
    if(v.is_number_integer() || v.is_number_float())
        return static_cast<float>(v.get<double>());
    if(v.is_string())
        return std::stof(v.get<std::string>());
    throw std::runtime_error(std::string("config: '") + key +
                            "' must be number or string, got " + v.type_name());
}

static Gpt2Config load_config(const std::string& path)
{
    std::ifstream f(path);
    if(!f.good())
    {
        throw std::runtime_error("Cannot open config: " + path);
    }
    json j = json::parse(f);

    Gpt2Config cfg;
    cfg.vocab_size           = config_get_int(j, "vocab_size", 50257);
    cfg.hidden_size          = config_get_int(j, "n_embd", 768);
    cfg.intermediate_size     = config_get_int(j, "n_inner", 4 * cfg.hidden_size);
    cfg.num_hidden_layers    = config_get_int(j, "n_layer", 12);
    cfg.num_attention_heads  = config_get_int(j, "n_head", 12);
    cfg.max_position_embeddings = config_get_int(j, "n_positions", 1024);
    cfg.layer_norm_eps       = config_get_float(j, "layer_norm_epsilon", 1e-5f);
    cfg.eos_token_id         = config_get_int(j, "eos_token_id", 50256);
    cfg.bos_token_id         = config_get_int(j, "bos_token_id", 50256);
    cfg.validate();
    return cfg;
}

// ── Weight cache ─────────────────────────────────────────────────────────

using WeightCache = std::map<std::string, std::vector<std::uint8_t>>;

static WeightCache load_weights_to_memory(const std::string& path)
{
    io::SafeTensorsReader reader(path);
    WeightCache cache;
    for(const auto& name : reader.tensor_names())
    {
        cache[name] = reader.read_tensor(name);
    }
    return cache;
}

static void apply_weight_cache(
    graph::module::Module& model,
    const WeightCache& cache)
{
    for(const auto& [name, tensor] : model.named_parameters_recursive())
    {
        auto it = cache.find(name);
        if(it != cache.end())
        {
            auto copy = it->second;
            tensor->data()->set_bind_hint(std::move(copy));
            tensor->mark_input(true);
        }
    }
}

// ── Python invocation ────────────────────────────────────────────────────

static bool run_python_conversion(const Args& args)
{
    std::vector<std::string> argv_strs = {
        "python3",
        "examples/gpt2_generate.py",
        "--model",
        args.model,
        "--output-dir",
        args.output_dir,
    };
    if(!args.prompt.empty())
    {
        argv_strs.push_back("--prompt");
        argv_strs.push_back(args.prompt);
    }

    std::vector<char*> argv;
    argv.reserve(argv_strs.size() + 1);
    for(auto& s : argv_strs)
        argv.push_back(s.data());
    argv.push_back(nullptr);

    std::cout << "Running: python3 examples/gpt2_generate.py --model "
              << args.model << " --output-dir " << args.output_dir;
    if(!args.prompt.empty())
        std::cout << " --prompt " << args.prompt;
    std::cout << "\n";

    pid_t pid = fork();
    if(pid < 0)
    {
        std::cerr << "fork() failed\n";
        return false;
    }
    if(pid == 0)
    {
        execvp("python3", argv.data());
        std::cerr << "execvp(python3) failed\n";
        _exit(127);
    }
    int status;
    if(waitpid(pid, &status, 0) != pid)
    {
        std::cerr << "waitpid() failed\n";
        return false;
    }
    return WIFEXITED(status) && WEXITSTATUS(status) == 0;
}

// ── Read prompt_ids.txt ──────────────────────────────────────────────────

static std::string read_file_trimmed(const std::string& path)
{
    std::ifstream f(path);
    if(!f.good())
        return {};
    std::string s((std::istreambuf_iterator<char>(f)),
                   std::istreambuf_iterator<char>());
    while(!s.empty() && (s.back() == '\n' || s.back() == '\r'))
        s.pop_back();
    return s;
}

// ── Greedy argmax over last-position logits ──────────────────────────────

static std::int64_t argmax_last_position(
    const std::vector<float>& logits,
    Index vocab_size,
    Index seq_len)
{
    auto offset = static_cast<std::size_t>((seq_len - 1) * vocab_size);
    auto begin = logits.begin()
        + static_cast<std::ptrdiff_t>(offset);
    auto end = begin + vocab_size;
    return static_cast<std::int64_t>(std::distance(
        begin, std::max_element(begin, end)));
}

// ── Main ─────────────────────────────────────────────────────────────────

int main(int argc, char** argv)
{
    Args args = parse_args(argc, argv);

    if(!args.model.empty())
    {
        if(!run_python_conversion(args))
        {
            std::cerr << "Python conversion failed.\n";
            return 1;
        }
        args.config_path  = args.output_dir + "/config.json";
        args.weights_path = args.output_dir + "/weights.safetensors";

        if(args.prompt_ids_str.empty())
        {
            std::string ids_file = args.output_dir + "/prompt_ids.txt";
            args.prompt_ids_str = read_file_trimmed(ids_file);
        }
    }

    if(args.config_path.empty() || args.weights_path.empty()
       || args.prompt_ids_str.empty())
    {
        std::cerr
            << "Error: provide either (--model + --prompt) or "
               "(--config + --weights + --prompt-ids).\n"
               "Run with --help for usage.\n";
        return 1;
    }

    Gpt2Config config = load_config(args.config_path);
    std::cout << "Config: hidden=" << config.hidden_size
              << "  layers=" << config.num_hidden_layers
              << "  heads=" << config.num_attention_heads
              << "  vocab=" << config.vocab_size << "\n";

    std::vector<std::int64_t> tokens = parse_ids(args.prompt_ids_str);
    if(tokens.empty())
    {
        std::cerr << "Error: prompt token list is empty.\n";
        return 1;
    }
    std::cout << "Prompt: " << tokens.size() << " tokens\n";

    std::cout << "Loading weights from " << args.weights_path << " ...\n";
    WeightCache weights = load_weights_to_memory(args.weights_path);
    std::cout << "Cached " << weights.size() << " parameter tensors\n";

    Context context(1, 1, 0, "/tmp/nntile_ooc", 16777216, 0,
                    "localhost", 5001, 0);

    std::cout << "\n--- Generating (greedy, no KV-cache) ---\n";
    for(int step = 0; step < args.max_tokens; ++step)
    {
        Index seq_len = static_cast<Index>(tokens.size());

        NNGraph graph("gpt2_step");

        auto* input_ids = graph.tensor(
            {seq_len, 1}, "input_ids", DataType::INT64, false);
        input_ids->mark_input(true);

        // position_ids: arange(0, seq_len) broadcast to (seq_len, 1)
        std::vector<std::int64_t> pos_ids(seq_len);
        for(Index i = 0; i < seq_len; ++i)
            pos_ids[i] = static_cast<std::int64_t>(i);
        auto* position_ids = graph.tensor(
            {seq_len, 1}, "position_ids", DataType::INT64, false);
        position_ids->mark_input(true);

        Gpt2Causal model(&graph, "model", config);
        auto* output = model.forward(input_ids, position_ids);
        output->mark_output(true);

        apply_weight_cache(model, weights);

        TensorGraph::Runtime runtime(graph.tensor_graph());
        runtime.compile();
        runtime.bind_data("input_ids", tokens);
        runtime.bind_data("position_ids", pos_ids);
        runtime.execute();
        runtime.wait();

        auto logits = runtime.get_output<float>(output->name());

        std::int64_t next_id = argmax_last_position(
            logits, config.vocab_size, seq_len);
        tokens.push_back(next_id);

        std::cout << "step " << (step + 1) << "  token=" << next_id << "\n";

        if(next_id == config.eos_token_id)
        {
            std::cout << "EOS reached.\n";
            break;
        }
    }

    std::cout << "\nGenerated token IDs:";
    for(auto id : tokens)
    {
        std::cout << " " << id;
    }
    std::cout << "\n";

    std::cout << "\nTo decode, run:\n"
              << "  python3 -c \"from transformers import AutoTokenizer; "
              << "t = AutoTokenizer.from_pretrained('<model-dir>'); "
              << "print(t.decode([";
    for(std::size_t i = 0; i < tokens.size(); ++i)
    {
        if(i > 0) std::cout << ",";
        std::cout << tokens[i];
    }
    std::cout << "]))\"\n";

    return 0;
}
